### ISSUE TITLE
coffeescript -> javascript ES6 (ES2015) ?

### ISSUE THREAD

**Issue by simevo** [buildbot/buildbot#3804]
Coffeescript complicates the build, adds dependencies and makes contributing harder.

Coffeescript does not appear in the [top 100 programming languages tracked by the TIOBE index](https://www.tiobe.com/tiobe-index/), nor in the 23 tracked by [PYPL](https://pypl.github.io/PYPL.html) and not in the 15 tracked by [github](https://octoverse.github.com/).

Other projects are migrating from CS to JS ES6, for example:
https://medium.com/adstage-engineering/our-journey-from-coffeescript-to-es6-af297b0990ac

According to:
```
find www | grep 'coffee$' | xargs wc | tail -n 1
```
buildbot contains about 10 kLOC of coffeescript (that's conservative as it also counts empty lines and comments...).

**Comment by tardyp** [comment]
Hi @simevo ,
I think coffeescript was a good choice 4 years ago (we started in 2013), but it is certainly not anymore, indeed.

As the blog post title suggests, migrating out of coffeescript will be a Journey.

I am all for that if people want to start that journey, and propose a proper plan. e.g.

- [ ] switch from guanlecoja build system to webpack
- [ ] re-write the plugins system in webpack (being compatible with the old guanlecoja based looks useful)
- [ ] convert the coffee code to es6 (with decaffeinate)
   - [ ] will probably mean to modify decaffeinate to understand ng-classify.

**Comment by simevo** [comment]
javascript is a "vibrant ecosystem" which also means that is risky: a new and fashionable dialect / tool may be abandoned in 3 years or less

buildbot is a server component and should be much stabler than that; this means that you should be extra careful when choosing new dependencies / tools

having said that, for the sake of packageability (for debian/ubuntu, see: https://bugs.debian.org/790241):
- getting rid of guanlecoja would be great
- webpack is fine
- bower is also deprecated but yarn is not yet in debian so it would be best to wait a bit on that

it'd be interesting to hear as many comments as possible from the other team members on this issue !

**Comment by dragon512** [comment]
I tend to find that typescript and web components ( like https://www.polymer-project.org/) have been good models for me in the Web space dev area. 
Typescript gives us some type safety and feature of what is next in Javascript es6 and es7 land today.
Web component allows us to make modular/component items that can be easily added to existing page. Communication happens in a standard way. 

I have found this gives me the best of what is new but in a safe flexible way

**Comment by tardyp** [comment]
@simevo

Packageability is another problem, as our release tarballs contain the javascript already bundled (like e.g. grafana).
We know that debian has the policy of requiring to build everything from source, but I don't think this is realistic, and we see other example (e.g. grafana) which skip that by shiping their app with bundled javascript. Grafana is 50% typescript (45kloc), 50% go (48kloc)
http://docs.buildbot.net/latest/relnotes/0.9.0.html#note-for-distro-package-maintainers-the-npm-dependency-hell

Re vibrant ecosystem: The problem is that we don't have a choice. For nine, we wanted a live updated web UI, so we needed javascript. 2013's Javascript (ES5) was not at all what es6 is now, so for me at that time, it was too difficult to maintain a decent size application on that langage. 

At coffeescript was a langage close enough to python, and could attract python developer.
With the experience now, I don't think this was a reality. Coffeescript has its own caveats and does not really reduce the ramp-up time for a python developer. 
I do think, with 4 years experience, that I am more productive using coffeescript that I am using ES5.
I am also convinced that I would get even more productivity boost with ES6 and the tooling of 2017 (VSCode + webpack hot-reload is amazing). But that requires a lot of work to switch.

Re typescript/polymer. For me those are not from me an industry standard, so if we choose to migrate to that, we will have in 3 years the same issue, another js based langage that has much better traction and ecosystem and documentation and stackoverflow database.
Switching UI framework is an even bigger work.

**Comment by dragon512** [comment]
These are just suggestions from my work at Intel trying to make stuff work in large teams. Typescript I would argue is a safe bet. I found that this worked better with large teams than CoffeeScript. I know people that just love CoffeeScript, however, I found that it was very liberal on return statement value and () which made it easy for people to make code that was hard to read and easy to break. I am not one of these python programmers that feel CoffeeScript should be suggested to them for Web work. Web component is risker I agree, but as a dev model using the HTTP event model, not some new random javascript event setup in some library that may or not work with this other library was very nice. However, I was able to make web pages easily quickly that worked on phones to desktops with it on evergreen browsers or better .. which should be the defacto standard as of today. I only suggest it as it would be a way to allow buildbot to have UI components that one can easily add to a random Dashboard

I think whatever you choose it would be nice if we can easily drop-in replacement like we use to be able to in buildbot 8. I would like to be able to extend the UI based on a well define Data model.

Again these are just suggestions... There are a lot of good options. The ones I suggest was the way I was able to get a team of people in different geo regions to work together in a "sane" way. Like you point out everything is always changing. The Web space seems very prone to the just rewrite the world effect.

**Comment by tardyp** [comment]
Drop-in replacement is already possible by design.
We actually had a proof of concept of that working with angular-material (stored inside md-base). However, it is not maintained, and didn't work last time I tried.

**Comment by uglycoyote** [comment]
I completely agree with @dragon512 that Typescript is the way to go here, especially if the buildbot UI is staying with Angular.  

I'm glad this topic has come up as I have also been feeling a lot of frustration with the coffeescript. I have a strong python background, and only only a modest amount of javascript experience, but despite coffeescript supposed similarity to python I personally found that coffeescript was not a benefit over Javascript.  The primary drawback was that being new to Angular, it made it really hard to compare sample code from the web to what was going on in buildbot, particularly with the use of ng-classify.  I understand how ng-classify was appealing, however it made it really hard for me to know which concepts were Angular vs ng-classify vs guanlecoja vs (other technology) concepts.

@tardyp makes the argument that Typescript are not industry standard and may be a passing fad, but I would counter with a couple of points there:   Firstly, my sense is that the Typescript team has done a lot of work to ensure that Typescript is on a converging path ECMAScript. So rather than being a whole distinct syntax/language, it's is basically ES6 plus type annotations.   Secondly, Typescript is no more a passing fad than Angular is.  I would point out that Angular 2 has completely bought in to Typescript (their sample code and tutorials use it and they recommend it), and buildbot has bought in to Angular (though not Angular 2 at this point).  I don't think it makes sense in the long term to stick with a dead version of Angular with serious performance issues. If upgrading to Angular 2, using Typescript seems like the obvious choice.   

I'm not necessarily a fan of blindly following the herd, actually I personally would prefer Typescript regardless of whether it was using Angular for the UI or some other solution.  The type annotations provide a huge productivity boost, for the simple reason that it makes it much easier to navigate in an unfamiliar codebase.  A big part of my difficulty with the current buildbot UI codebase is that there's no way to know where anything is defined without text searching, and most of my searches turn up way to many false positives. It's a huge chore to sift through and try to find the relevant results.  Particularly when buildbot code is referencing some external dependency, a text search either needs to include all of "node_modules" to find it (slow and too many false positives) or you need to search something like the bundled vendors.js file (where you lose the context of knowing where it came from).  With Typescript it's just "Go To Definition" and you are there (you usually land in the typings ".d.ts" file of the relevant library)

I recently went through the [Angular 2 Heroes tutorial](https://angular.io/tutorial) and also did a bit of playing around with hooking up an angular 2 app to bulidbot's REST API.  I was amazed at how good the experience was (using Sublime or VSCode to get the Typescript completion, Angular CLI for genereting nice clean and consistent folder layouts, etc.)

I have had positive experiences with webpack as well, and would welcome switching to it.

I realize that it's not going to happen overnight, but if there are some small steps we can take in this direction i would be happy to contribute if I can.

**Comment by tardyp** [comment]
@uglycoyote 
I think what you are saying is reasonable.

Switching to Angular 2 means that this will break compatibility with any custom plugin/dashboard that people might have written already. So basically those people will be stuck on old version of buildbot.

>I realize that it's not going to happen overnight, but if there are some small steps we can take in this direction i would be happy to contribute if I can.

The problem is that I don't see any small steps that is not preceded by a huge step :-/

**Comment by uglycoyote** [comment]
@tardyp wrote:
> The problem is that I don't see any small steps that is not preceded by a huge step :-/

The [angular upgrade document](https://angular.io/guide/upgrade) breaks it down in to small steps, it's not necessary to convert your entire application to Angular 2 at the same time.  They recommend having a hybrid AngularJS / Angular2 application where you can upgrade bits and pieces at a time until you no longer need AngularJS.   This may be a route where existing plugins can continue to work alongside upgraded Controllers/Services in the core.  As a small step, they actually recommend upgrading to Typescript first before even starting with the angular stuff.  Converting Coffee to Typescript (or ES6) could be done in as small increments as needed.

**Comment by tardyp** [comment]
> When you use ngUpgrade, what you're really doing is running both AngularJS and Angular at the same time. All Angular code is running in the Angular framework, and AngularJS code in the AngularJS framework. Both of these are the actual, fully featured versions of the frameworks. There is no emulation going on, so you can expect to have all the features and natural behavior of both frameworks.

I don't think we really want to ship a buildbot version with both version of angularjs running at the same time, so this would have to be done on a separate branch/separate repository.

I started to think about small steps, and one is to move the buildbot data module to es6.

I tried that in #3806 I was able to build something, but I couldn't run karma with that. If anybody with webpack experience can help, that would be nice.

**Comment by simevo** [comment]
I am glad to have sparked a lively discussion, sadly I can be of little help as I have not much to say on ES6/typescript or webpack.

@tardyp:
> Packageability is another problem, as our release tarballs contain the javascript already bundled (like e.g. grafana).
We know that debian has the policy of requiring to build everything from source, but I don't think this is realistic, and we see other example (e.g. grafana) which skip that by shiping their app with bundled javascript.

The plan is to redo your build process on debian infrastructure and with debianized tooling, starting from the github tarball as source (not the "output" tarball that you produce which for us is a "binary" artifact)

There is duplication of work but that's on the shoulders of the debian maintainer; agreeing on a simple build process and minimizing dependencies will make life much easier for her/him !

**Comment by rodrigc** [comment]
Hopefully we will be able to move away from coffeescript to ES6/typescript/webpack.
I think the original intent of coffeescript was to attract buildbot developers who were more experienced with Python than Javascript.

Javascript frameworks seem to come and go, and change very fast, so keeping up is difficult.
However, in the Javascript world, it looks like coffeescript isn't in the mainstream.

It would be good to use Javascript technologies that are more mainstream, and can attract more
Javascript developers.

**Comment by uglycoyote** [comment]
FYI, for anyone reading this and wanting to use plain JS for writing a new buildbot UI plugin:   it is of course possible to write your plugin in plain Javascript even though the rest of the framework uses coffeescript.  Though it was not immediately obvious how to get this working, so I thought I would share what I have discovered.

With the default setup of the Guanlecoja, gulp will automatically compile and bundle .coffee files but it will ignore javascript files that you add in your plugin's source directory.  The [DefaultConfig.coffee](https://github.com/buildbot/guanlecoja/blob/master/defaultconfig.coffee) file says 

```
# scripts (can be coffee or js)
        scripts: [
            'src/**/*.coffee'
            '!src/**/*.spec.coffee'
]
```

I'm not sure if it was intentional that the ```*.js``` was left out of this default configuration, since the comment above it says your scripts can be "coffee or js".  But you can override this configuration in your own plugin's  ```config.coffee``` by adding 

```
    files:
        # scripts (can be coffee or js)
        scripts: [
            'src/**/*.coffee'
            '!src/**/*.spec.coffee'
            'src/**/*.js'
        ]
```

And then it should start including your js sources into the final bundle.   There's no javascript transpilation going on though, so if you want to write ES6 then it may not work on older browsers.  Given that the audience is probably developers, this may be a non-issue.

Currently on plugin that I am working on (a modified version of the Console view, plus a bunch of dashboards that are specific to our company) I have somewhat abandoned Angular due to performance issues that I struggled but could not find a satisfactory way to get around (see [this issue](https://github.com/buildbot/buildbot/issues/3564), or [this StackOverflow post](https://stackoverflow.com/questions/47290376/strategies-for-angular-performance-in-large-table))

Although the rest of buildbot's UI is still in Angular, I am now using React to to the heavy lifting of rendering my Console's big table of information.  Essentially I just have a ```<div id="reactContent">``` in my angular template, and an Angular controller which gathers information from the buildbot ```dataService``` and then calls ReactDOM.render() and passes all of the information down into my React component.  

I needed to write JSX for my plugin which is a bit of an issue.  (for those who are not familiar, JSX is a javascript extension syntax that React uses which allows writing HTML tags inside the javascript code in a more natural way, but it requires transpilation to plain JS).  I tried to add support for JSX transpilation using Babel into guanlecoja's gulp configuration but so far have not been successful at that, so I have just been manually transpiling the small number of files that this are needed with a hotkey in my editor.  

This is all a bit of a Frankenstein mishmash of technologies, which I wouldn't necessarily advocate as it makes the overall picture even more murky, however in my case it seemed like the best path to get around my struggles with Angular.

**Comment by simevo** [comment]
Thanks for sharing. It's a mishmash, but offers a viable way to **start** a UI framework switch with a **small step** !

@uglycoyote, could you share a repo with a minimal scaffolding to create one such vanilla-js or react-js pliugin ? Even displaying "hello world" would suffice to yield a good starting point for anybody else trying something similar with other technologies/frameworks such as typescript, vue.js ...

**Comment by tardyp** [comment]
@uglycoyote Interesting approach. We can indeed start to write a boilerplate for react plugins with webpack bundling. This could allow us to assess the difference of performance for our usecases, and start experiment react.

All what you need to do for a plugin is to have some kind of build script (e.g webpack), which creates a scripts.js and a styles.css inside the static folder of your package.

Note that there is already support for starting webpack instead of gulp thanks to @skrchnavy via 7d2076ac8758e1810236b3e0a3dab339153ab8a2 for a web plugin, maybe he can share a boilerplate in a gist?

**Comment by uglycoyote** [comment]
@simevo, @tardyp I have set up a repo called [buildbot-react-plugin-boilerplate](https://github.com/uglycoyote/buildbot-react-plugin-boilerplate).  It's still half-baked but it is working.  It adds a new page to the buildbot UI where there is a list of recent revisions.  I'm having trouble getting it the .less/.css part of the build working, so the style rules don't work yet, and there are some other rough edges that I'll smooth over later.

**Comment by tardyp** [comment]
@uglycoyote nice work! pretty exiting!

**Comment by rodrigc** [comment]
The JavaScript world moves very fast, and it is hard to predict what is going to be
the dominant way to do things in JavaScript, and which frameworks will fall out of fashion.

I have been interacting a lot more with JavaScript developers these days in my work.
In the area where I live (San Francisco Bay Area), there are multiple programming schools and bootcamps that are churning out JavaScript developers.   Students in computer science programs in universities are getting more exposed to JavaScript via web developer courses.

From what I am seeing and with the people I have been talking to, at least now,
Angular and React are two popular frameworks that people are learning.  A lot of companies are also using them, or are in the process of moving to them.

Microsoft is behind TypeScript, and Angular (backed by Google) uses TypeScript.  So at least those technologies seem to have momentum behind them.

CoffeScript has its supporters, but in terms of mainstream web development, it doesn't seem to have the same momentum as these other technologies.

I think https://github.com/buildbot/buildbot/pull/3806 to move away from CoffeeScript is a good direction to go in.  Even if buildbot moved towards ES5/ES6, I think that would make it easier for JavaScript developers to get into the core buildbot UI and provide patches.

The work that @uglycoyote did with react is very cool.  For people that want to develop custom dashboards on top of buildbot, it is much easier to hire a JavaScript developer with React experience, and get them to use @uglycoyote's template.

**Comment by tardyp** [comment]
I have spent a little bit more thinking in converting to typescript.
The decaffeinate script is a good start, but the resulting code is imho quite ugly.

The thing is data-module is written as an angular1.x module, which I don't think there is any idea that can infer types from such imports. 
And with angular promises.. So there always will be hacks and complication to move promises and events from angular to react.

I think the proper path would be to rewrite data-module in typescript from scratch so that it can be imported as a standard es6 module.

This will allow to really use the type system of typescript, and then use any modern UI framwork
I started another PoC to see how hard that would be:
https://github.com/tardyp/buildbot-data/tree/master

I realized this is quite a bit of work, as I only converted a small module in 2 afternoons.

**Comment by uglycoyote** [comment]
It seems like the data module would be one of the trickier ones to make type annotations for.  Specifically I was looking at the way generateEndponits in the data service file works.  That is basically a piece of code that is adding new getter functions dynamically, but for type annotations to work you have to add them at compile time.

Had you thought about how you would tackle that part of it?  Would you just make that part less clever perhaps and have explicit functions in the code for each end point, returning a different type for each?

**Comment by tardyp** [comment]
@uglycoyote I was thinking about generating the code from the raml, without dynamic trick.

**Comment by uglycoyote** [comment]
@rodrigc @tardyp I wasn't quite sure how to offer help at the decaf effort on the data_module (hard to get past the Angular injection issues on the #3806 branch, and not familiar enough with how data_module works to offer help with @tardyp's newer [data-module](https://github.com/tardyp/buildbot-data/tree/master) effort)

Since my own heavily modified React-based fork of the Console plugin is still using gulp and some coffee (mixed with .jsx), I thought I would try to tackle the decaffeination of the console_view plugin, and the official version seemed like a simpler place to start than my modified version.  

I did some experiments on a branch to decaffeinate the console_view plugin and also switch it to using webpack to build.  This is working and was not too difficult, based on my previous experiences with making the [buildbot-react-plugin-boilerplate](https://github.com/uglycoyote/buildbot-react-plugin-boilerplate) sample.  

You can check out the commits on this branch here:
https://github.com/uglycoyote/buildbot/commits/webpack_console_view

I decaf'ed the ```.spec.coffee``` but I'm not sure how to run it since it normally just runs as part of the gulp build.   I'm not sure if it would make sense to run Karma as part of the webpack build or just as a separate script like an ```npm test``` defined in ```package.json``` which runs Karma.

It seems like @rodrigc  is already moving forward with pushing some decaffeinated sources into master, but I don't know if there's any consensus around what to do with guanlecoja/gulp, so it's probably premature for me to do a pull request with these changes making one plugin use webpack while the others do not.  

What do you think my next step should be?  Should I propose a pull request with the decaf changes only and then punt the webpack changes for later?

**Comment by rodrigc** [comment]
@uglycoyote I would recommend tidying up whatever you have as much as possible, and then filing a PR, with *[WIP]* in the title.  Then we can get multiple people to collaborate and push things along.

I decaffeinated the smoke tests, which were pretty simple and had no dependencies.

Decaffeinating the other components will be harder.

BTW, I'm not a JavaScript developer at all.  I've started getting into JS at my job, because I use the Angular Protractor framework for doing UI-based automation tests.  But I'm a novice at JS.
So hopefully we can pull in some more experienced folks to help.
